sesman/config.c

```diff
@@ -166,9 +172,44 @@ config_read_globals(int file, struct config_sesman *cf, struct list *param_n,
         cf->enable_user_wm = 0;
     }
 
-    if ('\0' == cf->default_wm[0])
+    if (cf->default_wm == 0)
+    {
+        cf->default_wm = g_strdup("startwm.sh");
+    }
+    else if (g_strlen(cf->default_wm) == 0)
+    {
+        g_free(cf->default_wm);
+        cf->default_wm = g_strdup("startwm.sh");
+    }
+    /* if default_wm doesn't begin with '/', it's a relative path to XRDP_CFG_PATH */
+    if (cf->default_wm[0] != '/')
+    {
+        length = sizeof(XRDP_CFG_PATH) + g_strlen(g_cfg->default_wm) + 1; /* '/' */
+        buf = (char *)g_malloc(length, 0);
+        g_sprintf(buf, "%s/%s", XRDP_CFG_PATH, g_cfg->default_wm);
+        g_free(g_cfg->default_wm);
+        g_cfg->default_wm = g_strdup(buf);
+        g_free(buf);
+    }
+
+    if (cf->reconnect_sh == 0)
+    {
+        cf->reconnect_sh = g_strdup("reconnectwm.sh");
+    }
+    else if (g_strlen(cf->reconnect_sh) == 0)
+    {
+        g_free(cf->reconnect_sh);
+        cf->reconnect_sh = g_strdup("reconnectwm.sh");
+    }
+    /* if reconnect_sh doesn't begin with '/', it's a relative path to XRDP_CFG_PATH */
+    if (cf->reconnect_sh[0] != '/')
     {
-        g_strncpy(cf->default_wm, "startwm.sh", 11);
+        length = sizeof(XRDP_CFG_PATH) + g_strlen(g_cfg->reconnect_sh) + 1; /* '/' */
```

Comment:
i think its better to mention the addition on null terminator explicitly.


BTW, if we were using `strlen`directly gcc would optimize it since it smart enough to detect its applied on a constant value. (but we are using our g_strlen wrapper :\)